### ISSUE TITLE
no-unmodified-loop-condition: Inconsistent behavior in calling function on loop condition var depending on whether condition checks member or not

### ISSUE THREAD

**Issue by platinumazure** [eslint/eslint#5738]
**What version of ESLint are you using?**

2.5.3

**What parser (default, Babel-ESLint, etc.) are you using?**

default

**Please show your full configuration:**

```json
{
    "rules": {
        "no-unmodified-loop-condition": "error"
    }
}
```

**What did you do? Please include the actual source code causing the issue.**

Example 1:
```js
for (var i = []; i != 0; i.push(0)) {
}
```

Example 2:
```js
for (var i = []; i.length === 0; i.push(0)) {
}
```

**What did you expect to happen?**

Both examples feature the loop condition being modified, via a function call on the object that has side effects. I would expect neither to result in a lint error.

**What actually happened? Please include the actual, raw output from ESLint.**

Example 1 (where the condition variable is directly compared) results in a lint error:
> 1:18 - 'i' is not modified in this loop (no-unmodified-loop-condition)

Example 2 (where a property of the condition variable is compared) has no errors.

--------

My take on this is as follows:

1. If we call a function on an object, it can have side effects on the object and should count as a modification.
  a. A direct comparison with a number or a string could be taking advantage of a `valueOf()` or `toString()` call.
1. It shouldn't matter whether we're checking the variable itself or a property of the variable when evaluating whether a condition is modified.

But perhaps there's an overarching design I'm not aware of which would explain the reasoning here?

Also, my original use case used Moment.js and is less stupid-looking than Example 1. I created my examples to prove that the library didn't matter.

Original use case:
```js
for (var weekCursor = weekStartDate.clone().add("weeks", 1); weekCursor <= weekEndDate; weekCursor.add("weeks", 1)) {
    that._addEventForWeek(weekCursor, event);
}
```

Where `weekCursor.add("weeks", 1)` is a function that has side-effects for the object on which it's called.

**Comment by mysticatea** [comment]
Honestly, I hadn't noticed that simple comparisons can have side effects.
I think we should describe `valueOf`, `toString`, and `[Symbol.toPrimitive]` as known limitations in the document of this rule.

No longer I can believe comparisons in JS. :(

```js
const i = {value: 0, valueOf() { return ++this.value; }};
while (i < 10) {
    console.log(i);
}

**Comment by platinumazure** [comment]
@mysticatea It's unfortunate that comparisons can have side-effects and I'm okay with documenting that `toString`/`valueOf`/`Symbol.toPrimitive` *with side effects* cannot be realistically covered by this rule.

However, my examples above (even the moment.js one) does not involve side-effects during the comparison. Rather, it involves a side-effect in the loop afterthought which is not being accounted for in the loop condition, but only when the loop condition compares the variable directly. It would be great to have at least an option for allowing a function call on the condition variable to count as a modification due to internal state changes.

**Comment by mysticatea** [comment]
Ah, sorry, my intent was that implicit conversions (`valueOf`, `toString`, ...) are known limitations.

I think, normally we don't expect to modify a reference by calling a method of the reference.
`foo.doSomething()` can modify a state of `foo`, but it doesn't modify `foo` itself generally.

`foo < 10` expression might be `foo.valueOf() < 10`. It's an implicit conversion. `no-unmodified-loop-condition` doesn't focus on properties and methods since these can have side effects. So if the rule handled implicit conversions, the rule must ignore all simple comparisons. I don't think it's nice.

**Comment by anseki** [comment]
Hi,
It seems to mistake parsing condition that includes calculation, also.
For example, ESLint say nothing to this code. it is correct behavior:

```js
var i, max = 10;
for (i = 0; i <= max; ++i) {
  //
}
```

But, to this code:

```js
var i, max = 10, add = 5;
for (i = 0; i <= max + add; ++i) {
  //
}
```

ESLint said:

```
[eslint] 'max' is not modified in this loop. (no-unmodified-loop-condition)
[eslint] 'add' is not modified in this loop. (no-unmodified-loop-condition)
```

Strange to say, it said about variables that are not `i`.

**Comment by mysticatea** [comment]
@anseki What version of ESLint do you use? I cannot reproduce it on the online demo.

![image](https://cloud.githubusercontent.com/assets/1937871/14850385/a0637192-0cb3-11e6-922a-08857212b6f3.png)

**Comment by anseki** [comment]
@mysticatea I'm sorry, I used v2.2.0 for the code above.
I tried that with v2.8.0, then those messages are not shown.
I am sorry for the lack of confirmation.

**Comment by platinumazure** [comment]
I think my original reported case is simply too much of an edge case for this rule, so I'm going to close this. (Also, there's been no activity for over two months.) Thanks for the discussion.
